### ISSUE TITLE
diffMethPerChr problem

### ISSUE THREAD

**Issue by ruchi19flowchart** [al2na/methylKit#200]
Dear Developer,
I'm not able to resolve below error, earlier it was running in the same way. Need help
> myDiff
methylDiffDB object with 226446 rows
--------------
         chr start end strand    pvalue    qvalue   meth.diff
1 CP027619.1   222 223      * 0.1891361 0.7370846  16.6183575
2 CP027619.1   232 233      * 0.4545804 0.9147658   7.5098814
3 CP027619.1   351 352      * 0.4194584 0.9095865 -12.6495726
4 CP027619.1   584 585      * 0.7377141 0.9585066   0.9219083
5 CP027619.1   585 586      * 0.7770980 0.9585066  -0.8696202
6 CP027619.1   651 652      * 0.6406141 0.9585066  -1.1282835
--------------
sample.ids: C H G F
destranded FALSE
assembly: human
context: CpG
treament: 1 1 0 1
resolution: base
dbtype: tabix

> diffMethPerChr(myDiff,plot=FALSE,qvalue.cutoff=0.01, meth.cutoff=10)
Error in fix.by(by.x, x) : 'by' must specify a uniquely valid column

**Comment by niahughes** [comment]
I have also had this issue - I tried to open an issue on the Google forums but I think it may have been caught in a spam filter. This seems to happen when there is one or more conditions (methylation state + chromosome) with no Cs or tiles meeting the threshold (e.g. 0 tiles hypermethylated on chromosome 2, at q < 0.01/meth.diff > 50). I tested this by increasing/decreasing the thresholds until the error appeared/resolved and converting to GRanges to take a look at how many tiles exist in each condition. 

For what it's worth Ruchi, I have had luck taking the source code for this command, unwrapping it from its function, and running it on its own. I couldn't tell you why it works in that case and not by default, but the output is easy to create right now if you're willing to go in and dig around a little.

**Comment by al2na** [comment]
We usually need a reproducible example to resolve these issues. niahughes's solution might work. Maybe there is a check in the original function needed if that issue is indeed about chromosomes having no Cs or tiles with the given threshold.

**Comment by niahughes** [comment]
After a little digging, I figured out what I changed to get the command to work in my modified code. I set the "all" flag on [the merge command on line 1350 of diffMeth.R](https://github.com/al2na/methylKit/blob/master/R/diffMeth.R#L1350) to TRUE, then converted NAs to 0s; otherwise, my function was just dropping any chromosomes with 0 in at least one condition entirely. This resulted in a different problem (only plotting the chromosomes with both significantly hypo and hypermethylated tiles). I haven't been able to replicate the fix.by error outside of the diffMethPerChr call itself, unfortunately. I've also had a hard time creating a minimal replicable example, though I'm still working on that.

Here's what I ended up with; it's mostly the same, except for the fix noted below (and the barplot is now through ggplot2).


```
diffList <- list(list, of, diffMeth, objects)
diffNames <- list("names", "of", "diffMeth", "objects")
names(diffList) <- diffNames

qvalue.cutoff <- 0.01
meth.cutoff <- 50

dMBC <- Map(function(x, y){
  print(y)
  x=getData(x)
  temp.hyper=x[x$qvalue < qvalue.cutoff & x$meth.diff >= meth.cutoff,]
  temp.hypo =x[x$qvalue < qvalue.cutoff & x$meth.diff <= -meth.cutoff,]
  
  dmc.hyper=100*nrow(temp.hyper)/nrow(x) # get percentages of hypo/ hyper
  dmc.hypo =100*nrow(temp.hypo )/nrow(x)
  
  all.hyper.hypo=data.frame(number.of.hypermethylated=nrow(temp.hyper),
                            percentage.of.hypermethylated=dmc.hyper,
                            number.of.hypomethylated=nrow(temp.hypo),
                            percentage.of.hypomethylated=dmc.hypo)
  
  dmc.hyper.chr=merge(as.data.frame(table(temp.hyper$chr)),
                      as.data.frame(table(x$chr)),by="Var1")
  dmc.hyper.chr=cbind(dmc.hyper.chr,
                      perc=100*dmc.hyper.chr[,2]/dmc.hyper.chr[,3])
  
  dmc.hypo.chr=merge(as.data.frame(table(temp.hypo$chr)),
                     as.data.frame(table(x$chr)),by="Var1")
  dmc.hypo.chr=cbind(dmc.hypo.chr,
                     perc=100*dmc.hypo.chr[,2]/dmc.hypo.chr[,3])
  
######################
#ALTERED LINES START HERE
  dmc.hyper.hypo=merge(dmc.hyper.chr[,c(1,2,4)],
                       dmc.hypo.chr[,c(1,2,4)],by="Var1", all=TRUE)
  dmc.hyper.hypo[is.na(dmc.hyper.hypo)] <- 0
#END
######################

  dmc.hyper.hypo=dmc.hyper.hypo[order(as.numeric(sub("chr","",dmc.hyper.hypo$Var1))),] # order the chromosomes
  
  names(dmc.hyper.hypo)=c("chr","number.of.hypermethylated",
                          "percentage.of.hypermethylated",
                          "number.of.hypomethylated",
                          "percentage.of.hypomethylated")
  te <- melt(dmc.hyper.hypo, id.vars = c("chr", "number.of.hypermethylated", "number.of.hypomethylated"))
  te$chr_f <- factor(te$chr, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
  
  plot <- ggplot(data=te, aes(x=chr_f, y=value, fill=variable)) + 
    geom_bar(stat="identity") + 
    labs(title=paste0("Percent DM tiles in ", sub(".*? ", "", y), " context, ", sub(" .*", "", y), " comparison"), 
         x="Chromosome", y="% DM tiles") +
    scale_fill_manual(name="Methylation\nlevel",labels = c("Hypermethylated", "Hypomethylated"), values=c("#2b83ba", "#d7191c")) + 
    ylim(0, 0.8) +
    coord_flip()
  
  png(paste0("/home/nia/file/path",gsub(" ", "", y, fixed = TRUE),"-dMBC.png"))
  print(plot)
  dev.off()
  
  
  return(dmc.hyper.hypo)
}, diffList, diffNames)
```

**Comment by niahughes** [comment]
I've developed a small reprex for this issue. The data is based on the test1 and control1 samples from the data included with mK but has some dummy methylation values added for two more chromosomes. I think I've determined that the output is based on two things: whether the data is an in-memory or database object, and whether the chromosome numbers start with the "chr" prefix or not.

If the "chr" prefix is present, regardless of whether the data is in-memory or not, the output is correct:

```
> diffMethPerChr(myDiff.chr.DB, qvalue.cutoff = 0.01, meth.cutoff = 51, plot=FALSE)
$diffMeth.per.chr
    chr number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1 chr21                        36                      2.142857                       39                     2.321429
2 chr22                         0                      0.000000                       29                    90.625000
3 chr23                         0                      0.000000                       29                    90.625000

$diffMeth.all
  number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1                        36                       2.06422                       97                     5.561927

> diffMethPerChr(myDiff.chr, qvalue.cutoff = 0.01, meth.cutoff = 51, plot=FALSE)
$diffMeth.per.chr
    chr number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1 chr21                        36                      2.142857                       39                     2.321429
2 chr22                         0                      0.000000                       29                    90.625000
3 chr23                         0                      0.000000                       29                    90.625000

$diffMeth.all
  number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1                        36                       2.06422                       97                     5.561927
```
However, if the prefix is not present and the data is in-memory, the output will only include chromosomes with significantly DM tiles in both conditions (i.e. it will drop chromosomes):

```
> diffMethPerChr(myDiff.nochr, qvalue.cutoff = 0.01, meth.cutoff = 51, plot=FALSE)
$diffMeth.per.chr
  chr number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1  21                        36                      2.142857                       39                     2.321429

$diffMeth.all
  number.of.hypermethylated percentage.of.hypermethylated number.of.hypomethylated percentage.of.hypomethylated
1                        36                       2.06422                       97                     5.561927
```

And if the prefix is not present and the data is in a tabix DB, the fix.by error occurs:

```
> diffMethPerChr(myDiff.nochr.DB, qvalue.cutoff = 0.01, meth.cutoff = 51, plot=FALSE)
Error in fix.by(by.x, x) : 'by' must specify a uniquely valid column
```

Here's the data:
[control-chr.myCpG.txt](https://github.com/al2na/methylKit/files/4637469/control-chr.myCpG.txt)
[control-nochr.myCpG.txt](https://github.com/al2na/methylKit/files/4637470/control-nochr.myCpG.txt)
[test-chr.myCpG.txt](https://github.com/al2na/methylKit/files/4637471/test-chr.myCpG.txt)
[test-nochr.myCpG.txt](https://github.com/al2na/methylKit/files/4637472/test-nochr.myCpG.txt)

And the script:
[DMBC-reprex.txt](https://github.com/al2na/methylKit/files/4637474/DMBC-reprex.txt)

**Comment by alexg9010** [comment]
Thanks niahughes for the investigations and the reprex!  
I will have a look at this bug as soon as I can, but for now nias suggestions   seems to be a good way to mitigate the error.

Best, 
Alex

**Comment by alexg9010** [comment]
Hi @niahughes and @ruchi19flowchart ,

A fix for this issue will be available in the next days with version 1.14.2 from Bioconductor. 

Please let me know if the issue still occours.

Best,
Alex
